whole-tale/gwvolman
# Bug Fixes & Progress Updates for Publishing

This PR is a work in progress-but any feedback is appreciated. In retrospect I should have split this into two PRs.

## gwvolman/dataone_metadata.py

### Set Custom Resource Map URL
I created `def createSimpleResourceMapCustom(ore_pid, scimeta_pid, sciobj_pid_list)` for the ability to add the URL to the ORE constructor (wasn't possible with `createSimpleResourceMap`). This was done to fix a bug where files weren't being properly located because the resource map was pointing to dataone production.


### Fix Access Rule
DataONE relies on the access rule to determine who can modify a published dataset. 
I fixed this at line 441 by adding the admin group. 

I noticed that the `sys_meta.submitter` wasn't being set, so I added it at line 422.

<br>
<br>

##  gwvolman/utils.py

### Added the dev cn to `is_in_network` and `is_dev_url`:
These methods should have been checking against the development coordinating node.

###  Added `extract_user_name`
This method takes a JWT token and returns the full name inside.

### Added `get_resource_map_user`:
There's an issue where ORCIDs can't be https in the resource map. I created this method to make the ORCID HTTP.

### Added `generate_dataone_guid` to fix #34 
DataONE like to have `urn:uuid` attached to the front of GUIDs.



<br>
<br>

## gwvolman/publish.py

### Use DataONE GUID Format
Once change that you'll see throughout the file is the replacement of the uuid calls with `generate_dataone_guid`.

### Tale Yaml Format Changes
These can be ignored for now, and are subject to change as Craig and I flesh out the format. (LN 99-335)

### Raise Exceptions on Failure
There are a few places where I modified how errors are handled. Instead of hiding them, I let pass them up to gwvolman when we're unable to continue. (`create_upload_repository` is going to go away when we introduce repo2docker).

### create_upload_remote_file
I added a method that takes a remote HTTP file, downloads it, and then uploads it to DataONE.


### Send Job Status Updates

In the main publish method I added calls to `job_manager.updateProgress` so that the UI can do things like update progressbars and let the user know the current status. 
<br>
<br>

## Other Notes

The following functions are related to the serialization format and may be depreciated
1. `create_metadata_info_structure`
2. `create_environment_structure`
3. `create_outer_structure`
4. `create_upload_tale_yaml`

`create_upload_repository` will be depreciated with repo2docker; I put a DEVNOTE in for it.

## To Do

### Hardcoded URL: #36 
Line 47 in gwvolman/dataone_metadata.py needs to be changed
Line 576 gwvolman/publish.py

### Correct MB Conversion
Line 663 in gwvolman/publish.py needs to get cleaned up (and do a proper conversion)

### Remove `logging.debug` statements
This commit should get squashed

### Organize commits

## Testing
This can be tested with the `dataone_publishing` branch on girder_wholetale.

1. Sign in to https://dev.nceas.ucsb.edu
2. Obtain your API key from account > my profile > Settings > Authentication Token
3. Navigate to your girder deployment and open the publish/dataone endpoint
4. Set `authToken` to the API key from above
5. Set `licenseSPDX` to `CC-BY-4.0`
6. Set `remoteMemberNode ` to `https://dev.nceas.ucsb.edu/knb/d1/mn`
7. Set `taleId` to the ID of any Tale
8. Set `itemIds` to a list of items ["item_id1", "item_id2", etc] that you want to publish
9. Click `Try it out!`
10. Navigate to `https://dev.nceas.ucsb.edu` and see that your Tale has been published
